Async API server/streams impl & functional tests + Allow to use local SRT source repo instead of remote

tchakabam opened the pull request:
... Will add more infos on this and tests around async API.

The work on the build script reflects a need to use patched SRT sources. Also some small improvements to build speed by detecting number of cores etc, as well as streamlining package installation.

birme commented on package.json:
```diff
@@ -4,13 +4,13 @@
-    "rebuild": "node-gyp rebuild",
-    "clean": "node-gyp clean",
+    "install": "npm run build-srt && npm run rebuild",
+    "build-srt": "node scripts/build-srt-sdk.js",
```
I might be blind, but when/where does it install the git-clone package now?

tchakabam replied:
it is runtime dependency of the whole package here, line 41 of this file here :)

-> it will get installed into `node_modules` as this package gets installed (calling `npm install` in root of this project, or if this package is a dependency of another package and gets installed along other deps in a tree.)

tchakabam replied:
normally you should not have to call `npm install <some-package>` explicitely ever here (only when you want to "manually" install a package or add/save it as dependency somewherere in a package/project context - calling the `npm install <some-package>` CLI  is a dev-time manual thing). here `git-clone` is a plain dev-dependency, and should be treated like all the others - there is no reason to make it more complicated than that.

the idea here is that you put all the runtime or dev-time deps into the appropriate list inside this file below (as you did for other deps already), and then once the package got installed properly you are ready to use any of them, assuming them in node_modules, or via any require/import module-loader which will resolve them there.

birme replied:
Ah yes of course. Is the git-clone added as a dev-dependency in this PR?

tchakabam replied:
actually i figure, `git-clone` here already was (before this PR), is and should be a plain dependency (not dev-dependency), so it is in the package dependencies section.

why is this a runtime dependency and not a devDependency, well we are getting confused here because we need these build-toolchain stuff (along with node-gyp/addon-api & friends) at runtime as people install this stuff in their env. dev-dependency is the toolchain we need to build/test things. But, and this is exceptional, we actually build the add-on JIT before runtime. and I think it is a good way to do it so, I think actually that is what you are supposed to do and all node-addons do (I am myself new to the wizardry of building add-ons in fact ... :)) I think it is needed to do so for x-platform and node-version ABI support otherwise one would need to ship a lot of bins at publish time :)

tchakabam commented on package.json:
```diff
@@ -4,13 +4,13 @@
+    "build-srt": "node scripts/build-srt-sdk.js",
+    "rebuild": "$(npm bin)/node-gyp rebuild -j $(echo \"console.log(require('os').cpus().length)\" | node)",
+    "clean": "$(npm bin)/node-gyp clean",
     "test": "$(npm bin)/jasmine",
```
@birme One question, why are we using `$(npm bin)/` here in the first place?

I have applied it consistently in the present patch, but realized it might no be needed at all.

If for example `jasmine` is installed in the closest node_modules bin path, the shell run by npm for this script will select it over any global installed executable (in case you wanted to ensure the local version is used).

You would rather have to enforce a global version to be used, but that would only make sense if that package was installed in the local project bin path at all.

birme replied:
I want to use the local node_modules version to not have to "pollute" the global node_modules space. For example `jasmine` is a dev-dependency

tchakabam replied:
Yes. And that means it gets installed in your `node_modules`(of the project). With that the case, as I tried to explain, the npm script shell will run the local bin without needing an explicit path (your *local* `node_modules/.bin` is prepended to the `PATH` var of that created shell in which the package script runs).

In short -> What you want will exactly work without the `npm bin` thing. Using the local package installed bin (no matter if dev-dep or runtime dep) is the default over any global bin path installs :) 

So, if that is what you want and it is the only requirement here, we can remove all `$(npm bin)/` here, because the behavior you want already is provided by putting there just `jasmine` or `eslint` (for example). npm will make sure local bins will resolve this before user/system-wide installs.

birme replied:
Aha, I understand. I didn't know that.

tchakabam replied:
No worries. Sorry if my explanation was confusing at first maybe. By "local" i was meaning the node_modules created in the project root here. Obviously there is also anything installed in /usr/local/bin and other places on the system that one could call "local", which I was meaning by "global" :) (because of `npm install -g`) ;)

So, yes anyhow, if you want to do something straight forward like this, dont think too complicated, those package scripts have been created exactly with that idea in mind to be able to launch the installed dependencies bins without having to specifiy a path :) npm doesn't do much really that is true, but that's one thing it definitely manages hehe :D

tchakabam replied:
It is funny that the `npm` docs are not very clear on this (but they aren't on many things): https://docs.npmjs.com/misc/scripts#path 

They say "added" to the path, which does not mean prepended, in the sense that it does not clarify order or resolution between those and other installs.

Here is a more historical and clear statement where this law of nature was introduced, `npm bin` output is prepending the shell `PATH` for lifecycle scripts since v1: https://nodejs.org/en/blog/npm/npm-1-0-global-vs-local-installation/#test-runners-and-stuff

```
npm makes ./node_modules/.bin the first entry in the PATH environment variable when running any lifecycle scripts, so this will work fine, even if your program is not globally installed:
```

birme replied:
As always very educating :)
Please go ahead and remove the usage of $(npm bin) in this PR if you want to.

tchakabam replied:
I am mostly stumbling upon my own doubts as I am trying to answer these questions :D

tchakabam commented:
@birme I pushed my latest WIP updates, but I am on vacation until the 14th fyi. We will continue on this after that. It turns out with what we needed to do it makes more sense to provide a server class using the async API like in the latest commits.

tchakabam commented on spec/async_srt_spec.js:
```diff
@@ -1,15 +1,6 @@
 const { SRT, AsyncSRT } = require('../index.js');
 
-describe("Async SRT Library", () => {
```
note to self: not sure why this was removed, should actually stay?

tchakabam replied:
was removed because duplicated in plain SRT spec

tchakabam commented on src/async-reader-writer.js:
```diff
@@ -0,0 +1,73 @@
+const DEFAULT_READ_BUFFER = READ_BUF_SIZE; // typical stream buffer size read in Node-JS internals
+
+/**
+ * @fires write
```
remove this

tchakabam commented on src/async-write-modes.js:
```diff
@@ -0,0 +1,197 @@
+ * Performance:
+ *
+ * What we call a "Yielding Loop" will attempt to use any task-slot available by the runtime
+ * for executing writes, while ensuring a limited number
```
"for queueing writes" - they get executed in the worker / task-runner.

tchakabam commented on src/srt-server.js:
```diff
@@ -0,0 +1,322 @@
+
+  /**
+   *
+   * @param {number} port socket port number
```
"port"

tchakabam commented on src/srt-server.js:
```diff
@@ -0,0 +1,322 @@
+  /**
+   *
+   * @param {number} port socket port number
+   * @param {string} address optional, default: '0.0.0.0'
```
"local address"

tchakabam committed:
replace stream module by improved version of readable/writable impl

tchakabam committed:
rm server.js

tchakabam committed:
async api improvments:
- better tracing of calls from worker back and forth
- fix transferrable handling to avoid copying buffers for r/w
- optional debug logs
- completed jsdocs annotations
- add dispose method
- add setLogLevel method (analoguous to added binding)

tchakabam committed:
node-srt C bindings:
- add SetLogLevel to get libSRT log output if desired
- add OK static member
- add #define EPOLL_EVENTS_NUM_MAX 1024
- improve error string thrown in Read (add that it comes from srt_recvmsg)
- improve error string thrown in Write (add that it comes from srt_sendmsg2)
- misc isofunctional improvements (var names) and comments

tchakabam committed:
add SRT logging related JS-side helper

tchakabam committed:
rewrite flat TypeScript decl files without "module" keyword

tchakabam committed:
add ts enum decl for all libSRT enums

tchakabam committed:
async-worker: enable using transferrable for zero-copy
+ allow better debugging (like in api/dispatcher side)
+ misc improvements on code quality

tchakabam committed:
add async-helpers: various functions to help dealing with transferrables
+ tracing calls to native bindings in debug output

tchakabam committed:
add async read/write modes functions + async-reader-writer class
- these will allow for performing high-level r/w operations conveniently
at optimum throughput for larger pieces of payload i.e list of packets.

tchakabam committed:
add srt-server and srt-connection (can manage multiple clients),
- based on async-api
- can be used with reader/writer (i.e the underlying modes)

tchakabam committed:
srt-server/connection typings

tchakabam committed:
async srt spec: add dispose method usage (but commented out as crashing atm)

tchakabam committed:
async srt spec: rm redundant checks on SRT static members (they are done
in other spec already)

tchakabam committed:
promises api spec: formal fixes

tchakabam committed:
stream spec: add dummy test

tchakabam committed:
package.json:
- put gyp toolchain in runtime deps (since the build happens on install)
- add JEST test runner
- shorten check-tsc script
- rebuild script: check & use all CPU cores available
- run rebuild actually on install, not preinstall (fixes deps not being there)
- remove preinstall and thus "npm install git-clone" in the package scripts

tchakabam committed:
update package lock

tchakabam committed:
update typings index not to need triple-slashs anymore

tchakabam committed:
in srt.ts example: check for read return value type

tchakabam committed:
build-srt-sdk script:
- allow to use any local libSRT code repo
- when using make: use all amount of cores available for build
- isolate better code running on different platforms

tchakabam committed:
update package main index with new things

tchakabam committed:
add enum typings index

tchakabam committed:
add jest config

tchakabam committed:
add "use strict" on async-srt-await example

tchakabam committed:
add integration/smoke testing for client-to-server one-way burst write

tchakabam committed:
readme: add note on build prerequisites

tchakabam committed:
readme: add infos on new components SRTServer/Connection & AsyncReaderWriter

birme reviewed (approved):
Thanks for a great contribution! Massive
